packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2524,6 +2525,11 @@ export class KeyringController extends BaseController<
     } catch (error) {
       console.error(error);
       this.#unsupportedKeyrings.push(serialized);
+      if (this.#keyringsMetadata.length > this.#keyrings.length) {
+        // There was already a metadata entry for the keyring, so
+        // we need to remove it
+        this.#keyringsMetadata.splice(this.#keyrings.length, 1);
+      }
```

Comment:
The downside of this solution is that if we consider the following scenario:
- There is some metadata for a keyring
- The keyring fails to restore, for any reason, and metadata will be removed
- The keyring is then supported again

The keyring will then have a different ID, compared to what it had initially. Unfortunately, to avoid this, we'd have to store the metadata along with its keyring in the vault directly.

Reply:
We have no mechanism today for moving a keyring from "unsupported" to "supported", to my knowledge. If and when we implement this, we can perhaps treat it like a new keyring and generate new metadata on-the-spot.

That should work OK.

Any other state relying on the metadata would get cleared when it was first moved to unsupported, that is one downside that would not avoid. But I think that's acceptable. Making a keyring unsupported must inevitably impact state that depends on that keyring.

Reply:
Oh, sorry I think I was mixing this up with `lostIdentities`. I see what you mean, thanks.

Reply:
I'm not sure this will work in that case. It breaks the assumption of a 1:1 match between keyrings and metadata.

I see you're trying to guard against this with the length check, but this doesn't work if we have two non-contiguous unsupported keyrings.

e.g. keyrings A, B, C, D where B is unsupported, it would remove metadata for B giving us A, C, D. Good so far.

But then if keyring D also becomes unsupported, then on-load we would encounter A, B, C, D again on unlock and would remove the 2nd metadata entry when we encounter B (not knowing B's metadata was already removed). That ends up reassigning the metadata entry for D to C.

Reply:
I'd recommend that we keep metadata for unsupported keyrings, but label it as unsupported somehow. In the same list so we preserve the order.

Reply:
That would also requiring _adding_ metadata for unsupported keyrings, if we find a case where the metadata doesn't exist yet.

Reply:
> That would also requiring adding metadata for unsupported keyrings, if we find a case where the metadata doesn't exist yet.

Given that after the vault is updated we would have `A, C, D, B` as keyring and `mA, mC, mD` as metadata, if B turns out to be supported again we would just add metadata for B with the current code, because B would be seen as "new"

Reply:
We synced on this and decided to go with the "metadata in vault" idea as being the least confusing and risky